## EDA: Examining a Dataset Drill

For this drill, we will using a modified dataset from the CO Animal Shelter info (named PACFA and can be found here: https://ag.colorado.gov/ics/pet-animal-care-facilities-act-pacfa/animal-shelter-and-rescue-individual-statistics )

The data consists of two separate files - one for the 2019 statistics and one for the 2020 statistics.  

Your boss wants you to answer a few of questions from the dataset:
- Overall numbers for 2019 and 2020 in these categories:
  - Intake for cats & dogs
  - Outake for cats & dogs
  - Comparison of 2020 vs 2019 both in numbers and a percent
  - Make this a table so you can present it to your boss & other stakeholders
- Did data quality improve from 2019 to 2020?
- Anecdotally, during 2020 it seemed like everyone was adopting new pets - does this show in the data?


Some considerations about the dataset:
- In 2020, the collection requirements and process changed in an attempt to improve data quality. This also caused the column names to be slightly different between years. 
- You can combine the files into one but this isn't necessary; both approaches have different problems to solve!

### Examine the dataset

Use as many cells below as you'd like to examine the two files:

- 2019 Individual Shelter and Rescue Report.csv
- 2020 Animal Shelter and Rescue Individual Report.csv

Some areas to check (feel free to check more items also!):

- Columns and column names 
- Datatypes of each of the columns/variables
- Min/Max/Mean statistics of each of those columns/variables
- Any NA values?

In [1]:
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None)

In [2]:
# first lets go through the 2019 data
as19 = pd.read_csv('2019 Individual Shelter and Rescue Report.csv')

In [3]:
print(len(as19))
as19.head()

355


Facility Name  \
0      2 Blondes All Breed Rescue   
1  2nd Chance Vizsla Rescue, Inc.   
2            4 Paws 4 Life Rescue   
3                  9 Lives Rescue   
4             Acadiana Animal Aid   

   Starting Animal Statistics - In Shelter Count as of 1/1/2019 - Dogs-Adult  \
0                                               59.0                           
1                                                0.0                           
2                                               26.0                           
3                                                0.0                           
4                                               17.0                           

   Starting Animal Statistics - In Shelter Count as of 1/1/2019 - Dogs-Juvenile  \
0                                               30.0                              
1                                                0.0                              
2                                               16.0                              
3                                                0.0                              
4                                               44.0                              

   Starting Animal Statistics - In Shelter Count as of 1/1/2019 - Cats-Adult  \
0                                                0.0                           
1                                                0.0                           
2                                                6.0                           
3                                                0.0                           
4                                               76.0                           

   Starting Animal Statistics - In Shelter Count as of 1/1/2019 - Cats-Juvenile  \
0                                                0.0                              
1                                                0.0                              
2                                                3.0                              
3                                                0.0                              
4                                                9.0                              

   Starting Animal Statistics - In Shelter Count as of 1/1/2019 - Birds  \
0                                                0.0                      
1                                                0.0                      
2                                                0.0                      
3                                                0.0                      
4                                                0.0                      

   Starting Animal Statistics - In Foster Care Count as 1/1/2019 - Dogs-Adult  \
0                                                0.0                            
1                                                8.0                            
2                                                0.0                            
3                                                0.0                            
4                                               32.0                            

   Starting Animal Statistics - In Foster Care Count as 1/1/2019 - Dogs-Juvenile  \
0                                                0.0                               
1                                                0.0                               
2                                                0.0                               
3                                                0.0                               
4                                               41.0                               

   Starting Animal Statistics - In Foster Care Count as 1/1/2019 - Cats-Adult  \
0                                                0.0                            
1                                                0.0                            
2                                                0.0                            
3                                               35.0                  

In [4]:
as19.columns

Index(['Facility Name',
       'Starting Animal Statistics - In Shelter Count as of 1/1/2019 - Dogs-Adult',
       'Starting Animal Statistics - In Shelter Count as of 1/1/2019 - Dogs-Juvenile',
       'Starting Animal Statistics - In Shelter Count as of 1/1/2019 - Cats-Adult',
       'Starting Animal Statistics - In Shelter Count as of 1/1/2019 - Cats-Juvenile',
       'Starting Animal Statistics - In Shelter Count as of 1/1/2019 - Birds',
       'Starting Animal Statistics - In Foster Care Count as 1/1/2019 - Dogs-Adult',
       'Starting Animal Statistics - In Foster Care Count as 1/1/2019 - Dogs-Juvenile',
       'Starting Animal Statistics - In Foster Care Count as 1/1/2019 - Cats-Adult',
       'Starting Animal Statistics - In Foster Care Count as 1/1/2019 - Cats-Juvenile',
       'Animal Intake Statistics - Stray - Dogs-Adult',
       'Animal Intake Statistics - Stray - Dogs-Juvenile',
       'Animal Intake Statistics - Stray - Cats-Adult',
       'Animal Intake Statistics - Str

It looks like there is an extra column for some reason - I want to look at whats happening there. Most of the column names are very long so might need to rename the ones we need to use.

In [5]:
as19['Unnamed: 54'].value_counts(dropna=False)

NaN    355
Name: Unnamed: 54, dtype: int64

So this column is holding no information - could be just how the data is stored with an extra unnamed column in the dataset. Probably can drop this one. 

In [6]:
as19.dtypes

Facility Name                                                                                              object
Starting Animal Statistics - In Shelter Count as of 1/1/2019 - Dogs-Adult                                 float64
Starting Animal Statistics - In Shelter Count as of 1/1/2019 - Dogs-Juvenile                              float64
Starting Animal Statistics - In Shelter Count as of 1/1/2019 - Cats-Adult                                 float64
Starting Animal Statistics - In Shelter Count as of 1/1/2019 - Cats-Juvenile                              float64
Starting Animal Statistics - In Shelter Count as of 1/1/2019 - Birds                                      float64
Starting Animal Statistics - In Foster Care Count as 1/1/2019 - Dogs-Adult                                float64
Starting Animal Statistics - In Foster Care Count as 1/1/2019 - Dogs-Juvenile                             float64
Starting Animal Statistics - In Foster Care Count as 1/1/2019 - Cats-Adult              

Looks like the datatypes for everything is float - not sure if there are fractions in there or nans but for now float is ok

In [7]:
as19.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 355 entries, 0 to 354
Data columns (total 55 columns):
 #   Column                                                                                                  Non-Null Count  Dtype  
---  ------                                                                                                  --------------  -----  
 0   Facility Name                                                                                           349 non-null    object 
 1   Starting Animal Statistics - In Shelter Count as of 1/1/2019 - Dogs-Adult                               349 non-null    float64
 2   Starting Animal Statistics - In Shelter Count as of 1/1/2019 - Dogs-Juvenile                            349 non-null    float64
 3   Starting Animal Statistics - In Shelter Count as of 1/1/2019 - Cats-Adult                               349 non-null    float64
 4   Starting Animal Statistics - In Shelter Count as of 1/1/2019 - Cats-Juvenile                

This dataset looks pretty clean and doesnt have any non-null values in all but the last column

In [8]:
as19.describe()

Starting Animal Statistics - In Shelter Count as of 1/1/2019 - Dogs-Adult  \
count                                         349.000000                           
mean                                           11.338109                           
std                                            39.311755                           
min                                             0.000000                           
25%                                             0.000000                           
50%                                             0.000000                           
75%                                             7.000000                           
max                                           608.000000                           

       Starting Animal Statistics - In Shelter Count as of 1/1/2019 - Dogs-Juvenile  \
count                                         349.000000                              
mean                                            2.971347                              
std                                            26.253172                              
min                                             0.000000                              
25%                                             0.000000                              
50%                                             0.000000                              
75%                                             0.000000                              
max                                           476.000000                              

       Starting Animal Statistics - In Shelter Count as of 1/1/2019 - Cats-Adult  \
count                                         349.000000                           
mean                                            7.352436                           
std                                            24.655538                           
min                                             0.000000                           
25%                                             0.000000                           
50%                                             0.000000                           
75%                                             0.000000                           
max                                           242.000000                           

       Starting Animal Statistics - In Shelter Count as of 1/1/2019 - Cats-Juvenile  \
count                                         349.000000                              
mean                                            1.398281                              
std                                             5.558890                              
min                                             0.000000                              
25%                                             0.000000                              
50%                                             0.000000                              
75%                                             0.000000                              
max                                            50.000000                              

       Starting Animal Statistics - In Shelter Count as of 1/1/2019 - Birds  \
count                                         349.000000                      
mean                                            1.868195                      
std                                            32.208549                      
min                                             0.000000                      
25%                                             0.000000                      
50%                                             0.000000                      
75%                                             0.000000                      
max                                           601.000000                      

       Starting Animal Statistics - In Foster Care Count as 1/1/2019 - Dogs-Adult  \
count                                         349.000000                            
mean                                       

The list of items that need to be done for the 2019 data:
- drop the column with no information
- figure out what columns we need in the analysis; drop the rest 
- rename the columns will we use for ease of access

Next let's look at the 2020 data!

In [9]:
as20 = pd.read_csv('2020 Animal Shelter and Rescue Individual Report.csv')
print(len(as20))
as20.head()

450


Facility Name  1/1/2020\n Adult Dogs\n In Shelter  \
0  2 Blondes All Breed Rescue, Inc.                                78.0   
1    2nd Chance Vizsla Rescue, Inc.                                 0.0   
2              4 Paws 4 Life Rescue                                 5.0   
3                    9 Lives Rescue                                 0.0   
4           A Friend of Jack Rescue                                 0.0   

   1/1/2020\n Adult Dogs\n In Foster Care  2020\n Adult Dogs\n Stray  \
0                                    47.0                        0.0   
1                                     0.0                        0.0   
2                                     0.0                        0.0   
3                                     0.0                        0.0   
4                                     0.0                        0.0   

   2020\n Adult Dogs\n Owner Relinquished  \
0                                     0.0   
1                                     0.0   
2                                    11.0   
3                                     0.0   
4                                     1.0   

   2020\n Adult Dogs Transfer from another Colorado Organization  \
0                                                0.0               
1                                                0.0               
2                                                0.0               
3                                                0.0               
4                                                0.0               

   2020\n Adult Dogs\n Transfer from Out of State  \
0                                           379.0   
1                                             0.0   
2                                           454.0   
3                                             0.0   
4                                           181.0   

   2020\n Adult Dogs\n Other Intake  2020\n Adult Dogs\n Adoption  \
0                              40.0                         498.0   
1                               0.0                           0.0   
2                               0.0                         459.0   
3                               0.0                           0.0   
4                               0.0                         175.0   

   2020\n Adult Dogs\n Returned to Owner (RTO)  \
0                                          0.0   
1                                          0.0   
2                                          0.0   
3                                          0.0   
4                                          0.0   

   2020\n Adult Dogs\n Transfer to another Colorado Organization  \
0                                                0.0               
1                                                0.0               
2                                                0.0               
3                                                0.0               
4                                                2.0               

   2020\n Adult Dogs\n Transfer to Out of State  \
0                                           1.0   
1                                           0.0   
2                                           6.0   
3                                           0.0   
4                                           0.0   

   2020\n Adult Dogs\n Other Transfer  1/1/2020\n Juvenile Dogs\n In Shelter  \
0                                 0.0                                    0.0   
1                                 0.0                                    0.0   
2                                 0.0                                   18.0   
3                                 0.0                                    0.0   
4                                 0.0                                    0.0   

   1/1/2020\n Juvenile Dogs\n In Foster Care  2020\n Juvenile Dogs\n Stray  \
0                                       56.0                           0.0   
1                                        0.0                          

From the first look - the columns are completely different and there are more rows in this dataset (more shelters were added in 2020 maybe - seems like a big increase?)

In [10]:
as20.columns

Index(['Facility Name', '1/1/2020\n Adult Dogs\n In Shelter',
       '1/1/2020\n Adult Dogs\n In Foster Care', '2020\n Adult Dogs\n Stray',
       '2020\n Adult Dogs\n Owner Relinquished',
       '2020\n Adult Dogs Transfer from another Colorado Organization',
       '2020\n Adult Dogs\n Transfer from Out of State',
       '2020\n Adult Dogs\n Other Intake', '2020\n Adult Dogs\n Adoption',
       '2020\n Adult Dogs\n Returned to Owner (RTO)',
       '2020\n Adult Dogs\n Transfer to another Colorado Organization',
       '2020\n Adult Dogs\n Transfer to Out of State',
       '2020\n Adult Dogs\n Other Transfer',
       '1/1/2020\n Juvenile Dogs\n In Shelter',
       '1/1/2020\n Juvenile Dogs\n In Foster Care',
       '2020\n Juvenile Dogs\n Stray',
       '2020\n Juvenile Dogs\n Owner Relinquished',
       '2020\n Juvenile Dogs\n Transfer from another Colorado Organization',
       '2020\n Juvenile Dogs\n Transfer from Out of State',
       '2020\n Juvenile Dogs\n Other Intake',
       

Since the column names are different - we will have to find the matching columns in 2019 to compare them to 2020. We also have an extra column here. Lets check that out first.

In [11]:
as20['Unnamed: 49'].value_counts(dropna=False)

NaN    450
Name: Unnamed: 49, dtype: int64

Again its all NaNs so we can drop it.

In [12]:
as20.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 450 entries, 0 to 449
Data columns (total 50 columns):
 #   Column                                                            Non-Null Count  Dtype  
---  ------                                                            --------------  -----  
 0   Facility Name                                                     356 non-null    object 
 1   1/1/2020
 Adult Dogs
 In Shelter                                  356 non-null    float64
 2   1/1/2020
 Adult Dogs
 In Foster Care                              356 non-null    float64
 3   2020
 Adult Dogs
 Stray                                           356 non-null    float64
 4   2020
 Adult Dogs
 Owner Relinquished                              356 non-null    float64
 5   2020
 Adult Dogs Transfer from another Colorado Organization      356 non-null    float64
 6   2020
 Adult Dogs
 Transfer from Out of State                      356 non-null    float64
 7   2020
 Adult Dogs
 Other Intake     

Looking at the info - the columns are float again so that matches the first dataset. But there are 450 entries and most columns have 356 nulls. I thought it was slighlty off that over 100 new shelters were added year over year (almost 25% increase) so I wonder what is happening here. Let's take a look at the bottom of the dataset.

In [13]:
as20.tail()

Facility Name  1/1/2020\n Adult Dogs\n In Shelter  \
445           NaN                                 NaN   
446           NaN                                 NaN   
447           NaN                                 NaN   
448           NaN                                 NaN   
449           NaN                                 NaN   

     1/1/2020\n Adult Dogs\n In Foster Care  2020\n Adult Dogs\n Stray  \
445                                     NaN                        NaN   
446                                     NaN                        NaN   
447                                     NaN                        NaN   
448                                     NaN                        NaN   
449                                     NaN                        NaN   

     2020\n Adult Dogs\n Owner Relinquished  \
445                                     NaN   
446                                     NaN   
447                                     NaN   
448                                     NaN   
449                                     NaN   

     2020\n Adult Dogs Transfer from another Colorado Organization  \
445                                                NaN               
446                                                NaN               
447                                                NaN               
448                                                NaN               
449                                                NaN               

     2020\n Adult Dogs\n Transfer from Out of State  \
445                                             NaN   
446                                             NaN   
447                                             NaN   
448                                             NaN   
449                                             NaN   

     2020\n Adult Dogs\n Other Intake  2020\n Adult Dogs\n Adoption  \
445                               NaN                           NaN   
446                               NaN                           NaN   
447                               NaN                           NaN   
448                               NaN                           NaN   
449                               NaN                           NaN   

     2020\n Adult Dogs\n Returned to Owner (RTO)  \
445                                          NaN   
446                                          NaN   
447                                          NaN   
448                                          NaN   
449                                          NaN   

     2020\n Adult Dogs\n Transfer to another Colorado Organization  \
445                                                NaN               
446                                                NaN               
447                                                NaN               
448                                                NaN               
449                                                NaN               

     2020\n Adult Dogs\n Transfer to Out of State  \
445                                           NaN   
446                                           NaN   
447                                           NaN   
448                                           NaN   
449                                           NaN   

     2020\n Adult Dogs\n Other Transfer  \
445                                 NaN   
446                                 NaN   
447                                 NaN   
448                                 NaN   
449                                 NaN   

     1/1/2020\n Juvenile Dogs\n In Shelter  \
445                                    NaN   
446                                    NaN   
447                                    NaN   
448                                    NaN   
449                                    NaN   

     1/1/2020\n Juvenile Dogs\n In Foster Care  2020\n Juvenile Dogs\n Stray  \
445                                        NaN                           NaN   
446  

So that explains the reason this dataset is longer. It looks like there are extra blank rows at the bottom of the dataset that need to be removed. We will have to filter these out.

In [14]:
as20.describe()

1/1/2020\n Adult Dogs\n In Shelter  \
count                          356.000000   
mean                             8.898876   
std                             20.634332   
min                              0.000000   
25%                              0.000000   
50%                              0.000000   
75%                              7.000000   
max                            143.000000   

       1/1/2020\n Adult Dogs\n In Foster Care  2020\n Adult Dogs\n Stray  \
count                              356.000000                 356.000000   
mean                                 5.620787                  66.991573   
std                                 13.077430                 268.876541   
min                                  0.000000                   0.000000   
25%                                  0.000000                   0.000000   
50%                                  0.000000                   0.000000   
75%                                  6.000000                   4.250000   
max                                129.000000                3000.000000   

       2020\n Adult Dogs\n Owner Relinquished  \
count                              356.000000   
mean                                37.772472   
std                                157.994389   
min                                  0.000000   
25%                                  0.000000   
50%                                  3.000000   
75%                                 17.000000   
max                               2301.000000   

       2020\n Adult Dogs Transfer from another Colorado Organization  \
count                                         356.000000               
mean                                           11.893258               
std                                            38.020526               
min                                             0.000000               
25%                                             0.000000               
50%                                             0.000000               
75%                                             7.000000               
max                                           409.000000               

       2020\n Adult Dogs\n Transfer from Out of State  \
count                                      356.000000   
mean                                        48.803371   
std                                        125.148253   
min                                          0.000000   
25%                                          0.000000   
50%                                          1.000000   
75%                                         26.750000   
max                                        975.000000   

       2020\n Adult Dogs\n Other Intake  2020\n Adult Dogs\n Adoption  \
count                        356.000000                    356.000000   
mean                          12.398876                    106.356742   
std                           56.195668                    270.175997   
min                            0.000000                      0.000000   
25%                            0.000000                      0.000000   
50%                            0.000000                     19.000000   
75%                            0.000000                     94.500000   
max                          590.000000                   2846.000000   

       2020\n Adult Dogs\n Returned to Owner (RTO)  \
count                                   356.000000   
mean                                     49.087079   
std                                     198.617080   
min                                       0.000000   
25%                                       0.000000   
50%                                       0.000000   
75%                                       1.000000   
max                                    2140.000000   

       2020\n Adult Dogs\n Transfer to another Colorado Organization  \
count                                         356.000000               
mean                  

These values seem reasonable - showing 356 values in each column so this reinforces that some blank rows need to be filtered out.

The list of items that need to be done for the 2020 data:
- drop the column with no information
- figure out what columns we need in the analysis and how they match to the 2019 data columns; drop the rest 
- rename the columns will we use for ease of access
- drop the rows with all NaNs

Total list of items to fix for the 2019/2020 data:
- Drop the columns with no information
- Figure out what columns we need in the analysis; drop the rest
- Need to see what columns in 2019 are the same as the 2020 data
- Rename the columns will we use for ease of access
- 2020 drop the rows that have all NaNs